src/storage/src/source/generator/key_value.rs

```diff
@@ -0,0 +1,584 @@
+                                records_staged: emitted,
+                            },
+                        )
+                        .await;
+                }
+                tokio::task::yield_now().await;
+            }
+
+            cap.downgrade(&MzOffset::from(key_value.snapshot_rounds));
+            key_value.snapshot_rounds
+        } else {
+            cap.downgrade(&resume_offset);
+            progress_cap.downgrade(&resume_offset);
+            resume_offset.offset
+        };
+
+        let mut local_partitions: BTreeMap<_, _> = (0..key_value.partitions)
+            .into_iter()
+            .filter_map(|p| {
+                config.responsible_for(p).then(|| {
+                    (
+                        p,
+                        UpdateProducer::new(
+                            p,
+                            key_value.partitions,
+                            key_value.keys,
+                            key_value.snapshot_rounds,
+                            key_value.quick_rounds,
+                            key_value.batch_size,
+                            key_value.seed,
+                            upper_offset,
+                            key_value.include_offset.is_some(),
+                        ),
+                    )
+                })
+            })
+            .collect();
+        if !local_partitions.is_empty()
+            && (key_value.update_rate.is_some() || key_value.quick_rounds > 0)
+        {
+            let mut interval = key_value.update_rate.map(tokio::time::interval);
+
+            loop {
+                if local_partitions.values().all(|si| si.finished_quick()) {
+                    if let Some(interval) = &mut interval {
+                        interval.tick().await;
+                    } else {
+                        break;
+                    }
+                }
+
+                for up in local_partitions.values_mut() {
+                    updates_buffer.clear();
+                    upper_offset = up.produce_batch(&mut updates_buffer, &mut value_buffer);
+                    data_output.give_container(&cap, &mut updates_buffer).await;
+                }
+                cap.downgrade(&MzOffset::from(upper_offset));
+                progress_cap.downgrade(&MzOffset::from(upper_offset));
+                tokio::task::yield_now().await;
+            }
+        }
+
+        std::future::pending::<()>().await;
+    });
+
+    let status = [HealthStatusMessage {
+        index: 0,
+        namespace: StatusNamespace::Generator,
+        update: HealthStatusUpdate::running(),
+    }]
+    .to_stream(scope);
+    let stats_stream = stats_stream.concat(&steady_state_stats_stream);
+
+    (
+        stream.as_collection(),
+        Some(progress_stream),
+        status,
+        stats_stream,
+        vec![button.press_on_drop(), stats_button],
+    )
+}
+
+/// An iterator that produces keys belonging to a partition.
+struct PartitionKeyIterator {
+    // The partition.
```

Comment:
Doc comments for the fields 
```suggestion
    /// The partition.
```